<?xml version="1.0"?>
<launch>

<!-- _________________________________________________________________________________ -->
<!--  CHANGE THIS DEPENDING ON WHICH SENSOR YOU WANT TO USE (UNCOMMENT THE ONE U NEED) -->

<!-- <arg name = "sensor_type_selected" default = "VLP16" /> -->
<!-- <arg name = "sensor_type_selected" default = "HDL32E" /> -->
<!-- <arg name = "sensor_type_selected" default = "OUSTEROS0" /> -->
<arg name = "sensor_type_selected" default = "TILT3D" />
<!-- <arg name = "sensor_type_selected" default = "TILT4D" /> -->

<!-- ____________ -->
<!--     VLP16    -->
<!--   check scan topic  -->
<node pkg= "cloud_factory" 		        type = "generic_model_data" 	name = "generic_model_data" 	output="screen" if = "$(eval sensor_type_selected == 'VLP16')" >
		<param name="sensorType"  		type="string" value="VLP16"/>
		<param name="laserDataTopic"  	type="string" value="puck_rear/velodyne_points"/>
		<param name="tfLaserDataFrameID" type="string" value="map"/>
		<param name="raysIncl"  		type="int" value="16"/> 				<!-- number of vertical rays  -->
		<param name="inclMin"  			type="double" 	value="-0.26180"/> 		<!---15.0° lowest inclination ray (vertical)  -->
        <!-- <param name="inclMin"  			type="double" 	value="-0.785398"/> 		-45°  -->
  		 <param name="inclMax"  	type="double" 	value="0.26180"/> 		<!--+15.0° highest inclination ray (vertical)  -->
		<!-- <param name="inclMax"  			type="double" 	value="3.92699"/>		+225° highest inclination ray (vertical) -->
		<param name="inclRes"  			type="double" 	value="0.0349066"/>		<!--   2.0° inclination resolution - angle between two vertical rays -->
		<param name="azimMin"  			type="double" 	value="0.0"/>			<!--   0.0° azimuth start angle (horizontal)   -->
		<param name="azimMax"  			type="double" 	value="6.28319"/>		<!-- 360.0° azimuth stop angle (horizontal)    -->
		<param name="azimRes"  			type="double" 	value="0.00349066"/>	<!--   0.2° azimuth resolution - angle between two horizontal rays   -->
        <param name="radiusCloud"  		type="double" 	value="0.01"/>
        <param name="radiusIncrement"  	type="double" 	value="0.0001"/>
</node>

<!-- ____________ -->
<!--     HDL32E   -->
<!--   check scan topic  -->
<node pkg= "cloud_factory" 		        type = "generic_model_data" 	name = "generic_model_data" 	output="screen" if = "$(eval sensor_type_selected == 'HDL32E')" >
		<param name="sensorType"  		type="string" value="HDL32E"/>
		<param name="laserDataTopic"  	type="string" value="puck_rear/velodyne_points"/>
		<param name="tfLaserDataFrameID"  		type="string" value="map"/>
		<param name="raysIncl"  		type="int" value="32"/> 				<!-- number of vertical rays  -->
		<param name="inclMin"  			type="double" 	value="-0.535292"/> 	<!-- -30.67° lowest inclination ray (vertical)  -->
		<param name="inclMax"  			type="double" 	value="0.186227"/>		<!-- +10.67° highest inclination ray (vertical) -->
		<param name="inclRes"  			type="double" 	value="0.0232129"/>		<!--   1.33° inclination resolution - angle between two vertical rays -->
		<param name="azimMin"  			type="double" 	value="0.0"/>			<!--    0.0° azimuth start angle (horizontal)   -->
		<param name="azimMax"  			type="double" 	value="6.28319"/>		<!--  360.0° azimuth stop angle (horizontal)    -->
		<param name="azimRes"  			type="double" 	value="0.002897"/>		<!--  0.166° azimuth resolution - angle between two horizontal rays   -->
        <param name="radiusCloud"  		type="double" 	value="0.01"/>
        <param name="radiusIncrement"  	type="double" 	value="0.00003"/>
</node>

<!-- ____________ -->
<!--   OUSTEROS0  -->
<!--   insert scan topic  -->
<node pkg= "cloud_factory" 		        type = "generic_model_data" 	name = "generic_model_data" 	output="screen" if = "$(eval sensor_type_selected == 'OUSTEROS0')" >
		<param name="sensorType"  		type="string" value="OUSTEROS0"/>
		<param name="laserDataTopic"  	type="string" value="puck_rear/velodyne_points"/>
		<param name="tfLaserDataFrameID"  		type="string" value="map"/>
		<param name="raysIncl"  		type="int" value="128"/> 				<!-- number of vertical rays  -->
		<param name="inclMin"  			type="double" 	value="-0.785398"/> 	<!-- -45.0° lowest inclination ray (vertical)  -->
		<param name="inclMax"  			type="double" 	value="0.785398"/>		<!-- +45.0° highest inclination ray (vertical) -->
		<param name="inclRes"  			type="double" 	value="0.012217"/>		<!--   0.7° inclination resolution - angle between two vertical rays -->
		<param name="azimMin"  			type="double" 	value="0.0"/>			<!--   0.0° azimuth start angle (horizontal)   -->
		<param name="azimMax"  			type="double" 	value="6.28319"/>		<!-- 360.0° azimuth stop angle (horizontal)    -->
		<param name="azimRes"  			type="double" 	value="0.0061087"/>		<!--  0.35° azimuth resolution - angle between two horizontal rays   -->
        <param name="radiusCloud"  		type="double" 	value="0.01"/>
        <param name="radiusIncrement"  	type="double" 	value="0.00002"/>
</node>

<!-- ____________ -->
<!--   TILT3D  -->
<!--   insert scan topic  -->
<node pkg= "cloud_factory" 		        type = "generic_model_data" 	name = "generic_model_data" 	output="screen" if = "$(eval sensor_type_selected == 'TILT3D')" >
		<param name="sensorType"  		type="string" value="TILT3D"/>
		<param name="laserDataTopic"  	type="string" value="puck_rear/velodyne_points"/>
		<param name="tfLaserDataFrameID"  		type="string" value="map"/>
		<param name="raysIncl"  		type="int" value="90"/> 					<!-- number of vertical rays = number of tilts = ( |_angleStart - _angleStop| ) / unit  -> check dyn reconf params of ohm_sensors/ohm_tilt_scanner_3d/cfg/ReconfigureScan.cfg -->
		<param name="inclMin"  			type="double" 	value="0.0"/> 			<!-- xx° lowest inclination ray (vertical) -> check _angleStart -->
		<param name="inclMax"  			type="double" 	value="1.570796327"/>			<!-- 90° for test - highest inclination ray (vertical) -> check _angleStop -->
		<param name="inclRes"  			type="double" 	value="0.0174533"/>			<!-- 1° for test - inclination resolution - angle between two vertical rays -> check unit in ohm_sensors/ohm_tilt_scanner/src/TiltScannerNode.cpp -->
		<param name="azimMin"  			type="double" 	value="-2.35619"/>			<!--   -135.0° azimuth start angle (horizontal)   -->
		<param name="azimMax"  			type="double" 	value="2.35619"/>		<!-- 135.0° azimuth stop angle (horizontal)    -->
		<param name="azimRes"  			type="double" 	value="0.0174533"/>		<!-- 1° for test azimuth resolution - angle between two horizontal rays   -->
        <param name="radiusCloud"  		type="double" 	value="0.8"/>
        <param name="radiusIncrement"  	type="double" 	value="0.0001"/>
</node>

<!-- ____________ -->
<!--   TILT4D  -->
<!--   insert scan topic  -->
<node pkg= "cloud_factory" 		        type = "generic_model_data" 	name = "generic_model_data" 	output="screen" if = "$(eval sensor_type_selected == 'TILT4D')" >
		<param name="sensorType"  		type="string" value="TILT4D"/>
		<param name="laserDataTopic"  	type="string" value="puck_rear/velodyne_points"/>
		<param name="tfLaserDataFrameID"  		type="string" value="map"/>
		<param name="raysIncl"  		type="int" value="180"/> 					<!-- number of vertical rays = number of tilts = ( |_angleStart - _angleStop| ) / unit  -> check dyn reconf params of ohm_sensors/ohm_tilt_scanner_3d/cfg/ReconfigureScan.cfg -->
		<param name="inclMin"  			type="double" 	value="0.0"/> 			<!-- xx° lowest inclination ray (vertical) -> check _angleStart -->
		<param name="inclMax"  			type="double" 	value="3.14"/>			<!-- 90° highest inclination ray (vertical) -> check _angleStop -->
		<param name="inclRes"  			type="double" 	value="0.0174533"/>			<!-- 1° for test inclination resolution - angle between two vertical rays -> check unit in ohm_sensors/ohm_tilt_scanner/src/TiltScannerNode.cpp -->
		<param name="azimMin"  			type="double" 	value="0.0"/>			<!--   0.0° azimuth start angle (horizontal)   -->
		<param name="azimMax"  			type="double" 	value="6.28319"/>		<!-- 236.0° azimuth stop angle (horizontal)    -->
		<param name="azimRes"  			type="double" 	value="0.0174533"/>		<!--  1° for test //0.33° azimuth resolution - angle between two horizontal rays   -->
        <param name="radiusCloud"  		type="double" 	value="1.0"/>
        <param name="radiusIncrement"  	type="double" 	value="0.00002"/>
</node>


</launch>
